thirdparty/DiskANN/src/math_utils.cpp

```diff
@@ -370,12 +381,13 @@ namespace kmeans {
     std::uniform_int_distribution<size_t> distribution(0, num_points - 1);
 
     size_t tmp_pivot;
-    for (size_t j = 0; j < num_centers; j++) {
-      tmp_pivot = distribution(generator);
-      if (std::find(picked.begin(), picked.end(), tmp_pivot) != picked.end())
-        continue;
-      picked.push_back(tmp_pivot);
-      std::memcpy(pivot_data + j * dim, data + tmp_pivot * dim,
+    for (int64_t j = num_points - num_centers; j < num_points; j++) {
+      tmp_pivot = std::uniform_int_distribution<size_t>(0, j)(generator);
```

Comment:
use set to replace vector is good, but can you explain a little bit about this change here?

Reply:
In the original code there's the probability that `picked.size()` is less than `num_centers` because of `continue` in the for loop.
I used an algorithm called [Bob Floyds sampling](https://math.stackexchange.com/questions/178690/whats-the-proof-of-correctness-for-robert-floyds-algorithm-for-selecting-a-sin) to sample `num_centers` points from data.

Reply:
I didn't fully get, what will happen if `picked.size()` is less then `num_centers`?

BTW there is another place use this picked logic in this file, do we need to make them sync to each other if this change is necessary?

Reply:
If `picked.size()` is less than `num_centers`, some data of `pivot_data` will be some useless random value. It'll cause problems in the subsequent kmeans code.

That place actually used another algorithm, so there's no problem.

Reply:
Aha, you are right. I read the wrong line in code. Good catch!

Reply:
BTW, this algo looks reasonable to me. But what's the benefit of using this instead of simply use 
```
if (picked.count(tmp_pivot)) {
    j--;
   continue;
}
```

Reply:
![20220818-001626](https://user-images.githubusercontent.com/44047980/185190788-1f40ef0c-72fa-4689-b988-1597f7d9e451.jpg)
There is no obvious difference in performance